pyo3-macros-backend/src/pyclass.rs

```diff
@@ -163,9 +168,20 @@ impl PyClassPyO3Options {
             };
         }
 
+        let python_version = pyo3_build_config::get().version;
+
         match option {
             PyClassPyO3Option::Crate(krate) => set_option!(krate),
-            PyClassPyO3Option::Dict(dict) => set_option!(dict),
+            PyClassPyO3Option::Dict(dict) => {
+                if python_version >= PY_3_9 || !is_abi3() {
+                    set_option!(dict);
+                } else {
+                    return Err(syn::Error::new(
+                        dict.span(),
+                        "`dict` requires >= python 3.9 with the `abi3` feature",
+                    ));
+                }
```

Comment:
I find it easer to read if we use `ensure_spanned!` macro to enforce a condition and error if unmet:

```suggestion
                ensure_spanned!(
                    python_version >= PY_3_9 || !is_abi3(),
                    dict.span() => "`dict` requires >= python 3.9 with the `abi3` feature"
                );
                set_option!(dict);
```